pilot/pkg/config/kube/gateway/conversion.go

```diff
@@ -263,36 +304,144 @@ func buildHTTPVirtualServices(obj config.Config, gateways map[parentKey]map[k8s.
+				if cfg := meshRoutes[ns][host]; cfg != nil {
+					// merge http routes
+					vs := cfg.Spec.(*istio.VirtualService)
+					vs.Http = append(vs.Http, httproutes...)
+					sortHTTPRoutes(vs.Http)
+				} else {
+					sortHTTPRoutes(httproutes)
+					// build one VS per namespace+host
+					meshRoutes[ns][host] = &config.Config{
+						Meta: config.Meta{
+							CreationTimestamp: obj.CreationTimestamp,
+							GroupVersionKind:  gvk.VirtualService,
+							Name:              fmt.Sprintf("%s-%d", name, count),
+							Annotations:       parentMeta(obj, nil),
+							Namespace:         ns,
+							Domain:            domain,
+						},
+						Spec: &istio.VirtualService{
+							Hosts:    []string{host},
+							Gateways: []string{gw},
+							Http:     httproutes,
+						},
+					}
+					count++
+				}
+			}
+			continue
+		}
+		// it's a gateway route
+		if _, f := gatewayRoutes[gw]; !f {
+			gatewayRoutes[gw] = make(map[string]*config.Config)
+		}
+		for _, host := range hosts {
+			if cfg := gatewayRoutes[gw][host]; cfg != nil {
+				// merge http routes
+				vs := cfg.Spec.(*istio.VirtualService)
+				vs.Http = append(vs.Http, httproutes...)
+				sortHTTPRoutes(vs.Http)
+			} else {
+				sortHTTPRoutes(httproutes)
+				// build one VS per gateway+host
+				gatewayRoutes[gw][host] = &config.Config{
+					Meta: config.Meta{
+						CreationTimestamp: obj.CreationTimestamp,
+						GroupVersionKind:  gvk.VirtualService,
+						Name:              fmt.Sprintf("%s-%d", name, count),
+						Annotations:       parentMeta(obj, nil),
+						Namespace:         ns,
+						Domain:            domain,
+					},
+					Spec: &istio.VirtualService{
+						Hosts:    []string{host},
+						Gateways: []string{gw},
+						Http:     httproutes,
+					},
+				}
+				count++
+			}
+		}
 	}
-	vsConfig := config.Config{
-		Meta: config.Meta{
-			CreationTimestamp: obj.CreationTimestamp,
-			GroupVersionKind:  gvk.VirtualService,
-			Name:              name,
-			Annotations:       parentMeta(obj, nil),
-			Namespace:         obj.Namespace,
-			Domain:            domain,
-		},
-		Spec: &istio.VirtualService{
-			Hosts:    hosts,
-			Gateways: gatewayNames,
-			Http:     httproutes,
-		},
+}
+
+// generateMatchName returns the generated HTTPMatchRequest name including the original path type
+// e.g.:
+// "default/foo-0-PathPrefix"
+// "default/foo-1-RegularExpression"
+func generateMatchName(httpRoute config.Config, ruleIdx int, typ k8s.PathMatchType) string {
+	return fmt.Sprintf("%s/%s-%d-%s", httpRoute.Namespace, httpRoute.Name, ruleIdx, string(typ))
+}
+
+func getOriginalPathType(match *istio.HTTPMatchRequest) k8s.PathMatchType {
+	i := strings.LastIndexByte(match.Name, '-')
+	return k8s.PathMatchType(match.Name[i+1:])
+}
+
+// sortHTTPRoutes sort generated vs routes to meet gateway-api requirements
+// see https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.HTTPRouteRule
+func sortHTTPRoutes(routes []*istio.HTTPRoute) {
```

Comment:
I think we may be missing the fallback part:

```
If ties still exist across multiple Routes, matching precedence MUST be determined in order of the following criteria, continuing on ties:

The oldest Route based on creation timestamp.
The Route appearing first in alphabetical order by “{namespace}/{name}”.
```

Also importantly is the CreationTimestamp of the generate VirtualService, as that is used in a number of places in the later XDS config generation - right now I guess its just the timestamp of the first one?

Reply:
> I think we may be missing the fallback part:

The routes [have already been sorted based on "{namespace}/{name}"](https://github.com/istio/istio/commit/1622e7f68148d90560c9011744177bb69b55f4b1#diff-a0e8831b6aefb0ef9b2cd269fcd726b26fd1104120950952b5217a7b104ba153R84-R92) before they are converted to VS

> Also importantly is the CreationTimestamp of the generate VirtualService, as that is used in a number of places in the later XDS config generation - right now I guess its just the timestamp of the first one?

Yes it's the first one

Reply:
First one so the oldest right? If so that make sense to me

Reply:
Yeah